### PR TITLE
Add Zookeeper Watcher

### PR DESCRIPTION
**Pull request by craigbr**
Also, add average round trip time, running average for last 5
round trips, and total number of round trips

### CODE REVIEW

File: main/zookeeper_watcher.go [hunk 1]
```diff
@@ -0,0 +1,207 @@
+//
```
Comment (mhoppal) [inline main/zookeeper_watcher.go:2]: // Copyright 2017 Hewlett Packard Enterprise Development LP

File: main/zookeeper_watcher.go [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+		log.Fatalf("Invalid HEALTH_CHECK_PATH, must not be empty")
+	}
+	if !strings.HasPrefix(healthCheckPath, "/") {
+		healthCheckPath = "/" + healthCheckPath
+	}
+	zookeeperServers := configuration.ZookeeperServers
+	if len(zookeeperServers) < 0 {
+		log.Fatalf("Invalid ZOOKEEPER_SERVERS, must not be empty")
+	}
+	prometheusEndpoint := configuration.PrometheusEndpoint
```
Comment (mhoppal) [inline main/zookeeper_watcher.go]: why do you check all other variables are set correctly besides this one?

File: main/zookeeper_watcher.go [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+		// Start prometheus endpoint
+		http.Handle("/metrics", promhttp.Handler())
+		log.Fatal(http.ListenAndServe(prometheusEndpoint, nil))
+	}()
+
+	var connection *zk.Conn
+	var eventChannel <-chan zk.Event
+	connection, eventChannel, err = connect(zookeeperServers)
+	if err != nil {
+		log.Fatalf("Connecting to Zookeeper failed: %s", err)
```
Comment (mhoppal) [inline main/zookeeper_watcher.go]: why exit why not emit a connect_status_metric?

File: main/zookeeper_watcher.go [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+	}
+
+	for true {
+		ev := <-eventChannel
+		log.Infof("Event = %s", ev)
+		if ev.State == zk.StateConnected {
+			log.Infof("Zookeeper state is connected")
```
Comment (mhoppal) [inline main/zookeeper_watcher.go:160]: and a connected state metric here!

File: README.md [hunk 5]
```diff
@@ -32,22 +40,50 @@ Several parameters can be specified using environment variables:
...
+| `HEALTH_CHECK_PATH`   | `zookeeper-health-check` | Path to use for health check read/writes  |
+| `ZOOKEEPER_SERVERS`   |`localhost`               | Zookeeper servers                         |
+| `PROMETHEUS_ENDPOINT` | `0.0.0.0:8080`           | Endpoint for Prometheus metrics           |
+| `WATCHER_PERIOD`      |`600`                     | How often to do a read/write cycle        |
+| `WATCHER_TIMEOUT`     |`60`                      | How long to wait for message read         |
+
 ## Metrics
 
-### Kafka Watcher
+### Zookeeper Watcher
```
Comment (matrixik) [inline README.md]: Kafka

File: README.md [hunk 6]
```diff
@@ -1,17 +1,25 @@
...
-cycle to determine the health of Kafka. The Kafka status and various other metrics are exposed as
+Each Watcher periodically writes a message and then reads it back. It uses this cycle to
+determine the health of what it is watching. The status and various other metrics are exposed as
 Prometheus metrics. If there is a single read or write failure, the status goes to WARNING. Once
 there have been two consecutive read or write failures, the status goes to ERROR.
 
+### Kafka Watcher
+
+The Kafka Watcher writes a message to a Kafka topic and then reads it back from the same
+topic. This ensures the entire round trip functionality of Zookeeper is functioning.
```
Comment (matrixik) [inline README.md]: of Kafka?

File: README.md [hunk 7]
```diff
@@ -32,22 +40,50 @@ Several parameters can be specified using environment variables:
...
+| `zookeeper_min_round_trip_time`   | `gauge`   | Minimum Round Trip Time in seconds                |
+| `zookeeper_read_failure_count`    | `counter` | Number of failures reading messages               |
+| `zookeeper_running_average_round_trip_time` | `gauge` | Running Average Round Trip Time in seconds for last 5 messages |
+| `zookeeper_watcher_status`        | `gauge`   | Status of watcher: -1 = NOT_STARTED, 0 = OK, 1 = WARNING, 2 = ERROR |
+| `zookeeper_write_failure_count`   | `counter` | Number of failures writing messages               |
+
 ## Install (using Docker)
 
-See github.com/monasca/monasca-docker/kafka-watcher
+See github.com/monasca/monasca-docker/kafka-watcher and github.com/monasca/monasca-docker/zookeeper-watcher
```
Comment (matrixik) [inline README.md]: Both of this links will show 404 error on github

File: main/zookeeper_watcher.go [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+
+func main() {
+	configuration := watcherConfiguration{}
+	err := configEnv.Parse(&configuration)
+	if err != nil {
+		log.Fatalf("Failed to parse envrionment variables. %s", err)
+	}
+
+	healthCheckPath := configuration.HealthCheckPath
+	if len(healthCheckPath) < 0 {
```
Comment (matrixik) [inline main/zookeeper_watcher.go]: `< 1` or `<= 0`

File: main/zookeeper_watcher.go [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+		log.Fatalf("Invalid HEALTH_CHECK_PATH, must not be empty")
+	}
+	if !strings.HasPrefix(healthCheckPath, "/") {
+		healthCheckPath = "/" + healthCheckPath
+	}
+	zookeeperServers := configuration.ZookeeperServers
+	if len(zookeeperServers) < 0 {
```
Comment (matrixik) [inline main/zookeeper_watcher.go]: `< 1` or `<= 0`

File: watcher/watcher.go [hunk 10]
```diff
@@ -104,29 +113,48 @@ func CreateWatcher(watchable Watchable, period time.Duration, maxFailures int, t
...
+			Name: prefix + "running_average_round_trip_time",
+			Help: fmt.Sprintf("Running Average Round Trip Time in seconds for last %d messages",
+				RUNNING_AVERAGE_SLOTS)})
 	prometheus.MustRegister(writeFailuresMetric)
 	prometheus.MustRegister(readFailuresMetric)
 	prometheus.MustRegister(statusMetric)
 	prometheus.MustRegister(droppedMessagesMetric)
 	prometheus.MustRegister(minRoundTripTimeMetric)
 	prometheus.MustRegister(maxRoundTripTimeMetric)
+	prometheus.MustRegister(averageRoundTripTimeMetric)
```
Comment (matrixik) [inline watcher/watcher.go]: Register this in the same order like you initialized them in lines 116-128.

**Review by matrixik — CHANGES_REQUESTED** [review]

File: main/zookeeper_watcher.go (hunk 1 above)
Comment (craigbr) [reply main/zookeeper_watcher.go:2]: Done

File: main/zookeeper_watcher.go (hunk 2 above)
Comment (craigbr) [reply main/zookeeper_watcher.go]: Done

File: README.md (hunk 6 above)
Comment (craigbr) [reply README.md]: Done

File: README.md (hunk 5 above)
Comment (craigbr) [reply README.md]: Done

File: README.md (hunk 7 above)
Comment (craigbr) [reply README.md]: Done

File: main/zookeeper_watcher.go (hunk 9 above)
Comment (craigbr) [reply main/zookeeper_watcher.go]: Should be == 0. Definitely not < 0. Done

File: watcher/watcher.go (hunk 10 above)
Comment (craigbr) [reply watcher/watcher.go]: Good idea. Done

**Comment by craigbr** [comment]
@matrixik I've addressed all of your comments. Thanks for the review. Can you remove the "Changes requested"? Thanks

**Review by matrixik — APPROVED** [review]

**Pull request merged**